src/main/common/filter.c

```diff
@@ -168,151 +142,118 @@ float filterGetNotchQ(float centerFreq, float cutoffFreq)
 
     /* shift x1 to x2, input to x1 */
-    filter->x2 = filter->x1;
-    filter->x1 = input;
+    s->s[biquad_x2] = s->s[biquad_x1];
+    s->s[biquad_x1] = input;
 
     /* shift y1 to y2, result to y1 */
-    filter->y2 = filter->y1;
-    filter->y1 = result;
+    s->s[biquad_y2] = s->s[biquad_y1];
+    s->s[biquad_y1] = result;
 
     return result;
 }
 
-/* Computes a biquadFilter_t filter in df1 and crossfades input with output */
-FAST_CODE float biquadFilterApplyDF1Weighted(biquadFilter_t* filter, float input)
-{
-    // compute result
-    const float result = biquadFilterApplyDF1(filter, input);
-
-    // crossfading of input and output to turn filter on/off gradually
-    return filter->weight * result + (1 - filter->weight) * input;
-}
-
 /* Computes a biquadFilter_t filter in direct form 2 on a sample (higher precision but can't handle changes in coefficients */
-FAST_CODE float biquadFilterApply(biquadFilter_t *filter, float input)
+FAST_CODE float biquadFilterApply_cs(const FILTER_COEFFS_TYPE(biquad) *c, FILTER_STATE_TYPE(biquad) *s, float input)
 {
-    const float result = filter->b0 * input + filter->x1;
+    const float result = c->c[biquad_b0] * input + s->s[biquad_s1];
 
-    filter->x1 = filter->b1 * input - filter->a1 * result + filter->x2;
-    filter->x2 = filter->b2 * input - filter->a2 * result;
+    s->s[biquad_s1] = c->c[biquad_b1] * input - c->c[biquad_a1] * result + s->s[biquad_s2];
+    s->s[biquad_s2] = c->c[biquad_b2] * input - c->c[biquad_a2] * result;
 
     return result;
 }
 
-// Phase Compensator (Lead-Lag-Compensator)
 
-void phaseCompInit(phaseComp_t *filter, const float centerFreqHz, const float centerPhaseDeg, const uint32_t looptimeUs)
-{
-    phaseCompUpdate(filter, centerFreqHz, centerPhaseDeg, looptimeUs);
+DEFINE_FILTER_FN(biquad, DEFAULT_FILTER_DIMS);
+DEFINE_FILTER_FN2(biquad, biquadDF1, DEFAULT_FILTER_DIMS);
 
-    filter->x1 = 0.0f;
-    filter->y1 = 0.0f;
-}
+// Phase Compensator (Lead-Lag-Compensator)
 
-FAST_CODE void phaseCompUpdate(phaseComp_t *filter, const float centerFreqHz, const float centerPhaseDeg, const uint32_t looptimeUs)
+FAST_CODE void phaseCompFilterCoeffsCenPh(FILTER_COEFFS_TYPE(phaseComp) *c, const float centerFreqHz, const float centerPhaseDeg, const float dT)
 {
-    const float omega = 2.0f * M_PIf * centerFreqHz * looptimeUs * 1e-6f;
+    const float omega = 2.0f * M_PIf * centerFreqHz * dT;
     const float sn = sin_approx(centerPhaseDeg * RAD);
     const float gain = (1 + sn) / (1 - sn);
     const float alpha = (12 - sq(omega)) / (6 * omega * sqrtf(gain));  // approximate prewarping (series expansion)
 
-    filter->b0 = 1 + alpha * gain;
-    filter->b1 = 2 - filter->b0;
-    filter->a1 = 1 - alpha;
-
     const float a0 = 1 / (1 + alpha);
 
-    filter->b0 *= a0;
-    filter->b1 *= a0;
-    filter->a1 *= a0;
+    c->c[phaseComp_b0] = a0 * (1 + alpha * gain);
+    c->c[phaseComp_b1] = a0 * (2 - c->c[phaseComp_b0]);
+    c->c[phaseComp_a1] = a0 * (1 - alpha);
```

Comment:
Order of application is important here. line 237 is incorrect now. `c->c[phaseComp_b0]` is now multiplied by `a0`